### PR TITLE
Make plotting with gradients more fun

### PR DESCRIPTION
**Pull request by NilsNemitz**
Inspired by #1712, I looked at what options we have for plotting with colors according to values.

It turns out that QPen lets us do that without problem, but currently it is necessary to manually define a gradient. As a user, it is also not obvious that the coordinates used in the definition of the gradient directly correspond to the plotted values.

This PR adds convenience methods to get a `QBrush` or `QPen` from a ColorMap object.
It extends the functionality of the already existing `getGradient()` method to handle repeating and mirrored mapping modes.

It also adds a `reverse()` method to ColorMap, which reverses an existing map, which is often convenient for imported maps.

An example is included based on the discussion of #1737, demonstrating that smoothly updating plots are possible even with fancy coloring options.

### CODE REVIEW

File: examples/ColorGradientPlots.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+        self.traces = (
+            {'crv': curve0, 'buf': np.zeros( length ), 'ptr':0, 'ds': DataSource( signal_period=0.55 ) },
+            {'crv': curve1, 'buf': np.zeros( length ), 'ptr':0, 'ds': DataSource( signal_period=0.61, negative_period=0.55 ) },
+            {'crv': curve2, 'buf': np.zeros( length ), 'ptr':0, 'ds': DataSource( signal_period=0.65 ) },
+            {'crv': curve3, 'buf': np.zeros( length ), 'ptr':0, 'ds': DataSource( signal_period=0.52 ) },
+        )
+        self.timer = QtCore.QTimer(timerType=QtCore.Qt.PreciseTimer)
+        self.timer.timeout.connect(self.update)
+        timestamp = time.time()
```
Comment (j9ac9k) [inline examples/ColorGradientPlots.py:98]: Probably should use `time.perf_counter()` for float type, or `time.perf_counter_ns()`  for int type (in ns).  Also we have [ptime.py](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/ptime.py) but I'm not sure it's actually relevant any more or not...

**Comment by j9ac9k** [comment]
I just ran the example, this is slick!

The one suggestion I would have is to add a title to each plot with a short description of how it is determined what color to apply.

**Comment by NilsNemitz** [comment]
Dear @j9ac9k , I have added text labels to each plot to detail what they show. I have also switched to the performance counter, although that does not matter much as we are already averaging over a few frames to get a good number.

I also got Sphinx to work on my computer and did a full pass over the color map documentation.
Let me know what you think!

File: doc/source/colormap.rst [hunk 2]
```diff
@@ -1,8 +1,92 @@
...
+Sources for color maps
+----------------------
+
+Color maps can be user defined by assigning a number of *stops* over the range of 0 to 1. A color is given for each stop, and the in-between values are generated by interpolation.
+
+When map colors directly represent values, an improperly designed map can obscure detail over certain ranges of values, while creating false detail in others. PyQtGraph includes the perceptually uniform color maps provided by the `colorcet project` <https://colorcet.holoviz.org/>. Color maps can also be imported from the `colorcet` library or from `Matplotlib`, if either of these is installed.
+
+To see all available color maps, please run the ColorMap demonstration available in the suite of :ref:`examples`.
+
+Examples
```
Comment (ixjlyons) [inline doc/source/colormap.rst:23]: I'm a bit on the fence on this. I'm absolutely in favor of having small usage examples with visuals in the docs, but I think we should be careful about how to do it in a way that's scalable and won't lead to docs that are out of sync with the library. I've been working on an extension to [sphinx-gallery](https://sphinx-gallery.github.io/stable/index.html) which could be used to generate a gallery of the full example scripts. I've also dreamt of a sphinx extension to generate renderings of code blocks in the docs. I suppose we can always leave this in and replace it later if something like that ever gets going. Currently waiting on readthedocs to support installation of apt packages to install xvfb for rendering.

One thing we could do for now is use your scripts for generating these images and just list that code here via [`literalinclude`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude) and appropriate line numbers. It might be worth converting the code to a more "script-like" structure in that case.

File: pyqtgraph/colormap.py [hunk 3]
```diff
@@ -47,16 +55,25 @@ def listMaps(source=None):
...
+    
+    Parameters
+    ----------
+    name: str
+        Name of color map. In addition to the included maps, this can also 
+        be a path to a file in the local folder. See the files in the
+        `pyqtgraph/colors/maps/` folder for examples of the format.
+    source: str, optional
+        If omitted, a locally stored map is returned. Otherwise
+        - 'matplotlib' imports a map defined by Matplotlib.
```
Comment (ixjlyons) [inline pyqtgraph/colormap.py:71]: If there's no blank line between bullet points, they don't get rendered correctly. I'd just remove the bullet points and write it as a single sentence.

File: pyqtgraph/colormap.py [hunk 4]
```diff
@@ -47,16 +55,25 @@ def listMaps(source=None):
...
+
+    Returns a ColorMap object from a local definition or imported from another library.
+    The generated ColorMap objects are cached for fast repeated access.
+    
+    Parameters
+    ----------
+    name: str
+        Name of color map. In addition to the included maps, this can also 
+        be a path to a file in the local folder. See the files in the
+        `pyqtgraph/colors/maps/` folder for examples of the format.
```
Comment (ixjlyons) [inline pyqtgraph/colormap.py:68]: Double backticks for RST, annoying I know

**Review by ixjlyons — COMMENTED** [review]
Very nice demo. I've left some minor feedback mostly regarding RST formatting.

File: pyqtgraph/colormap.py [hunk 5]
```diff
@@ -180,22 +197,23 @@ def getFromColorcet(name):
...
 
-    The colors for intermediate values are determined by interpolating between 
+    The colors for intermediate values are determined by interpolating between
     the two nearest colors in RGB color space.
+    
+    A ColorMap object provides access to the interpolated colors by indexing with a float value:
+    `cm[0.5]` returns a QColor corresponding to the center of ColorMap `cm`.
 
-    To provide user-defined color mappings, see :class:`GradientWidget <pyqtgraph.GradientWidget>`.
+    To provide interactively user-defined color mappings, see :class:`GradientWidget <pyqtgraph.GradientWidget>`.
```
Comment (j9ac9k) [inline pyqtgraph/colormap.py:215]: a nice shorthand here would be 

```
:class:`~pyqtgraph.GradientWidget`
```

File: pyqtgraph/colormap.py [hunk 6]
```diff
@@ -330,25 +398,107 @@ def mapToFloat(self, data):
...
+        """
+        Returns a `QtGui.QLinearGradient` object corresponding to the ColorMap.
+        The span and orientiation is given by two points in plot coordinates.
+        
+        When no parameters are given for `p1` and `p2`, the gradient is mapped to the 
+        `y` coordinates 0 to 1, unless the color map is defined for a more limited range.
+        
+        Parameters
+        ----------
+        p1: QtCore.QpointF, default (0,0)
```
Comment (j9ac9k) [inline pyqtgraph/colormap.py:414]: super nit-picky, I know but should be `QtCore.QPointF`

File: pyqtgraph/colormap.py (hunk 3 above)
Comment (NilsNemitz) [reply pyqtgraph/colormap.py:72]: I added the missing blank lines, since I found this hard to read when it mixes the 'matplotlib' and 'colorcet' explanations in the same line.

File: pyqtgraph/colormap.py (hunk 5 above)
Comment (NilsNemitz) [reply pyqtgraph/colormap.py:218]: ⇒ changed.

File: pyqtgraph/colormap.py (hunk 4 above)
Comment (NilsNemitz) [reply pyqtgraph/colormap.py:68]: ⇒ changed.

File: doc/source/colormap.rst [hunk 7]
```diff
@@ -1,8 +1,92 @@
...
+that are colored according to their values along the horizontal or vertical axis.
+
+
+
+Sources for color maps
+----------------------
+
+Color maps can be user defined by assigning a number of *stops* over the range of 0 to 1. A color is given for each stop, and the in-between values are generated by interpolation.
+
+When map colors directly represent values, an improperly designed map can obscure detail over certain ranges of values, while creating false detail in others. PyQtGraph includes the perceptually uniform color maps provided by the `colorcet project` <https://colorcet.holoviz.org/>. Color maps can also be imported from the `colorcet` library or from `Matplotlib`, if either of these is installed.
```
[the remark this answers is not in the document]
Comment (NilsNemitz) [reply doc/source/colormap.rst]: ⇒ changed.
Colorcet doesn't seem to quite agree on whether the name is capitalized or not :)

File: doc/source/colormap.rst [hunk 8]
```diff
@@ -1,8 +1,92 @@
-ColorMap
-========
+Color maps
+==========
+
+A color map defines a relationship between scalar data values and a range of colors. 
+Color maps are commonly used to generate false color images, color scatter-plot points, 
+and illustrate the height of surface plots.
+
+PyQtGraph's ColorMap object can also be used a convenient source of colors from a
```
[the remark this answers is not in the document]
Comment (NilsNemitz) [reply doc/source/colormap.rst]: ⇒ changed.
I think a good approach is to have the first reference to `ColorMap` appear as the class reference, and the just use it as plain text through the rest of the page. Do we have other guidelines for that yet?

File: doc/source/colormap.rst [hunk 9]
```diff
@@ -1,8 +1,92 @@
-ColorMap
-========
+Color maps
```
[the remark this answers is not in the document]
Comment (NilsNemitz) [reply doc/source/colormap.rst]: ⇒ changed.

File: pyqtgraph/colormap.py (hunk 6 above)
Comment (NilsNemitz) [reply pyqtgraph/colormap.py]: ⇒ changed.

**Comment by NilsNemitz** [comment]
Dear @ixjlyons and @j9ac9k , thank you for the crash course in rst formatting. I am still making things up as I go along. I hope things work reasonably well now.

I have managed to directly pull the example code from the source files that generate the images. There is a convenient `dedent` parameter that takes out the extra spaces not needed in the examples.

I hope we'll get working support for the sphinx gallery at some point. When your library does graphics, it seems helpful to have images in the documentation!

Until it is a little clearer how this will work in practice, I'd suggest that we try to collect all the code that generates example images. This is in the images folder now, but that's only because I wasn't sure what the best location would be.

For the current sphinx format, a width of 420 pixels seems to work reasonably well for a half-width image, and that keeps the files small enough (17kB and 36 kB here) to just throw them into git for now; at least as long as we don't update them all the time :)

File: pyqtgraph/colormap.py [hunk 10]
```diff
@@ -115,7 +135,7 @@ def _getFromFile(name):
     return cm
 
 def getFromMatplotlib(name):
-    """ import colormap from matplotlib definition """
+    """ Generates a ColorMap object from a Matplotlib definition. Same as ``colormap.get(name, source='matplotlib')``. """
```
Comment (j9ac9k) [inline pyqtgraph/colormap.py:138]: should this perhaps be 

```
:method:~`pyqtgraph.colormap.get`
```

that way we get links within the docs?

Comment (j9ac9k) [reply pyqtgraph/colormap.py:138]: actually, scratch that, you're demonstrating an equivalent function, not looking to reference one...

File: pyqtgraph/colormap.py [hunk 11]
```diff
@@ -221,29 +239,28 @@ class ColorMap(object):
...
-                            :func:`mkColor() <pyqtgraph.mkColor>`.
-        mapping             Mapping mode (ColorMap.CLIP, REPEAT, MIRROR or DIVERGING)
-                            controlling mapping of relative index to color. String representations
-                            'clip', 'repeat', 'mirror' or 'diverging' are also accepted.
-                            CLIP maps colors to [0.0;1.0] and is the default.
-                            REPEAT maps colors to repeating intervals [0.0;1.0];[1.0-2.0],...
-                            MIRROR maps colors to [0.0;-1.0] and [0.0;+1.0] identically
-                            DIVERGING maps colors to [-1.0;+1.0]
-        ===============     =======================================================================
+        __init__(pos, color, mapping=ColorMap.CLIP)
```
Comment (j9ac9k) [inline pyqtgraph/colormap.py:242]: just out of curiosity, why is this line needed?  I see the output renders fine; but I'm just curious.

File: pyqtgraph/colormap.py [hunk 12]
```diff
@@ -221,29 +239,28 @@ class ColorMap(object):
...
+        
+        Parameters
+        ----------
+        pos: array_like of float in range 0 to 1
+            Assigned positions of specified colors
+        color: array_like of colors
+            List of colors, interpreted via 
+            :func:`mkColor() <pyqtgraph.mkColor>`.
+        mapping: str or int, optional
+            Controls
```
Comment (j9ac9k) [inline pyqtgraph/colormap.py:252]: get rid of newline?

File: pyqtgraph/colormap.py (hunk 11 above)
Comment (NilsNemitz) [reply pyqtgraph/colormap.py:242]: It makes it so that `ColorMap.CLIP` is spelled out just like that. The automatic signature shows the numerical value, and `mapping=2` is not so helpful.

File: pyqtgraph/colormap.py (hunk 12 above)
Comment (NilsNemitz) [reply pyqtgraph/colormap.py:252]: Probably. 
Sphinx has some kind of obscure method to determine line breaks in the parameter definitions, though.

With the bonus newline, this renders as 
mapping _(str or int, optional)_ - Controls [text continues]

Without the bonus newline, I _think_ this rendered as
mapping _(str or int, optional)_ -
Controls [text continues]

I find the hanging dashes ugly...
But I agree that it is probably not worth fighting sphinx over.

⇒ will remove

File: examples/ColorGradientPlots.py (hunk 1 above)
Comment (NilsNemitz) [reply examples/ColorGradientPlots.py]: (already updated)

**Comment by j9ac9k** [comment]
Should have merged this days ago, sorry for the wait @NilsNemitz This LGTM.  Thanks for jumping on this and taking the opportunity from the now-closed issue.
